RFCs/0031-Multi-Test-Run-Finalization.md

```diff
@@ -0,0 +1,304 @@
+    /// </summary>
+    public interface IDataCollectorAttachmentProcessor
+    {
+        /// <summary>
+        /// Gets the attachments Uris, which are handled by processor
+        /// </summary>
+        IEnumerable<Uri> GetExtensionUris();
+
+        /// <summary>
+        /// Indicates whether ProcessAttachmentSets is associative (e.g. f([a, b, c]) = f([f([a, b]), c]) = f([a, f([b, c])]))
+        /// </summary>
+        bool IsAssociative  { get; }
+
+        /// <summary>
+        /// Reprocess attachments generated by independent test executions
+        /// </summary>
+        /// <param name="attachments">Attachments to be processed</param>
+        /// <param name="progressReporter">Progress reporter. Accepts integers from 0 to 100</param>
+        /// <param name="logger">Message logger</param>
+        /// <param name="cancellationToken">Cancellation token</param>
+        /// <returns>Attachments after reprocessing</returns>
+        ICollection<AttachmentSet> ProcessAttachmentSets(ICollection<AttachmentSet> attachments, IProgress<int> progressReporter, IMessageLogger logger, CancellationToken cancellationToken);
+    }
+}
+```
+
+Method `GetExtensionUris` should provide all Uris for data collector attachments which are handled by current processor. Test platform will provide to processor only data collector attachments with such Uris. Result of method `ProcessAttachmentSets` shouyld contain only data collector attachments with such Uris.
+
+`IsAssociative` should indicate if `ProcessAttachmentSets` is [associative](https://en.wikipedia.org/wiki/Associative_property).
+
+If `IsAssociative` is `True` Test Platform may try to speed up whole process by reprocessing data collector attachments as soon as possible when any two test executions are done. For example let's assume we have 5 test executions which are generating 5 data collector attachments: `a1`, `a2`, `a3`, `a4` and `a5`. Test platform could perform invocations:
+* `var result1 = ProcessAttachmentSets([a1, a2, a3], ...);` when first 3 executions are done
+* `var result2 = ProcessAttachmentSets(result1.Concat([a4]), ...);` when 4th execution is done
+* `var finalResult = ProcessAttachmentSets(result2.Concat([a5]), ...);` when last test execution is done
+
+If `IsAssociative` is `False` then Test Platform will wait for all test executions to finish and call `ProcessAttachmentSets` only once:
+* `var finalResult = ProcessAttachmentSets([a1, a2, a3, a4, a5], ...);`
+
+By default `IsAssociative` should be `False`, unless processing can take longer time and it's beneficial to start the process as soon as possible.
+
+
+
+2. Introduce a new `FinalizeMultiTestRunAsync` method in [IVsTestConsoleWrapper](https://github.com/microsoft/vstest/blob/master/src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/IVsTestConsoleWrapper.cs) interface:
+
+```
+/// <summary>
+/// Provides back all attachments to TestPlatform for additional processing (for example merging)
```

Comment:
it would be nice to write somewhere the example of usage. that when you RunTests with a ITestRunEventsHandler, then the attachments coming from multiple HandleTestRunComplete invocations can be merged.

it took me a while for this method can be used

Reply:
Added this comment to another task that we will extend this documenation